klib/firewall.c

```diff
@@ -0,0 +1,601 @@
+}
+
+static int firewall_filter(struct pbuf *pbuf, struct netif *input_netif)
+{
+    list_foreach(&firewall.rules, elem) {
+        firewall_rule rule = struct_from_list(elem, firewall_rule, l);
+        if (firewall_match(pbuf, rule)) {
+            if (rule->drop)
+                goto drop_pkt;
+            break;
+        }
+    }
+    return 1;
+  drop_pkt:
+    pbuf_free(pbuf);
+    return 0;
+}
+
+static firewall_constraint_val firewall_create_constraint_val(heap h, value v, int val_size)
+{
+    if (!is_string(v) || !buffer_length(v))
+        return 0;
+    boolean neq;
+    if (peek_char(v) == '!') {
+        neq = true;
+        buffer_consume(v, 1);
+    } else {
+        neq = false;
+    }
+    u64 val;
+    if (!parse_int(v, 10, &val))
+        return 0;
+    firewall_constraint_val c = allocate(h, sizeof(*c));
+    assert(c != INVALID_ADDRESS);
+    c->c.equals = !neq;
+    switch (val_size) {
+    case 1:
+        c->val = val;
+        break;
+    case 2:
+        c->val = lwip_htons(val);
+        break;
+    case 4:
+        c->val = lwip_htonl(val);
+        break;
+    default:
+        c->val = lwip_htonl(val);
+        c->val = (c->val << 32) | lwip_htonl(val >> 32);
+        break;
+    }
+    return c;
+}
+
+static void firewall_dealloc_l3_constraint(heap h, firewall_constraint c)
+{
+    switch (c->type) {
+    case FW_L3_SRC: {
+        firewall_constraint_buf c_buf = (firewall_constraint_buf)c;
+        deallocate(h, c_buf, sizeof(*c_buf) + c_buf->len);
+        break;
+    }
+    case FW_L3_PROTO: {
+        firewall_constraint_val c_val = (firewall_constraint_val)c;
+        deallocate(h, c_val, sizeof(*c_val));
+        break;
+    }
+    }
+}
+
+static boolean firewall_rule_parse_ip4(heap h, firewall_rule rule, value spec)
+{
+    if (!is_tuple(spec)) {
+        rprintf("firewall: invalid ip value '%v'\n", spec);
+        return false;
+    }
+    rule->l3_match = allocate_vector(h, 2);
+    assert(rule->l3_match != INVALID_ADDRESS);
+    string src = get(spec, sym(src));
+    if (src) {
+        if (!is_string(src) || !buffer_length(src)) {
+            rprintf("firewall: invalid ip src value '%v'\n", src);
+            return false;
+        }
+        char *src_str = buffer_to_cstring(src);
+        boolean neq;
+        if (src_str[0] == '!') {
+            neq = true;
+            src_str++;
+        } else {
+            neq = false;
+        }
+        char *slash = runtime_strchr(src_str, '/');
+        if (slash)
+            *slash = '\0';
+        ip4_addr_t addr;
+        if (!ip4addr_aton(src_str, &addr)) {
+            rprintf("firewall: invalid ip address '%s'\n", src_str);
+            return false;
+        }
+        u64 netmask;
+        if (slash) {
+            buffer netmask_str = alloca_wrap_cstring(slash + 1);
+            if (!parse_int(netmask_str, 10, &netmask) || buffer_length(netmask_str) ||
+                (netmask == 0) || (netmask > 32) || (netmask & 0x7)) {
```

Comment:
I had put the constraint for the netmask to be a multiple of 8 bits so that the comparison between addresses could be as simple as a runtime_memcmp() call. But indeed this could be too much of a limitation, so I changed the code to support any netmask values from 1 up to the IP address size. The constraint buf still contains just be part of the address that is of interest for the comparison, so that there is no need for a buf length field in addition to the netmask field.